### PR TITLE
Fix writing ETags on singletons and nav props

### PR DESCRIPTION
**Pull request by mikepizzo**
### Issues
*This pull request fixes issue #926*  
*This pull request fixes issue #927*  

### Description
*Fixes writing of ETags on singletons and containment nav props*
*Ensures that null values are written in a delta payload*

### Checklist (Uncheck if it is not completed)
- [ x ] Test cases added
- [ x ] Build and test with one-click build and test script passed

### Additional work necessary
*None needed*

### CODE REVIEW

File: OData/src/System.Web.OData/OData/ETagMessageHandler.cs [hunk 1]
```diff
@@ -115,12 +115,11 @@ internal Task<HttpResponseMessage> SendAsync(HttpRequestMessage request)
...
         {
             IEdmModel model = entityInstanceContext.EdmModel;
-            IEdmEntitySet entitySet = entityInstanceContext.NavigationSource as IEdmEntitySet;
 
             IEnumerable<IEdmStructuralProperty> concurrencyProperties;
-            if (model != null && entitySet != null)
+            if (model != null && entityInstanceContext.NavigationSource != null)
             {
-                concurrencyProperties = model.GetConcurrencyProperties(entitySet).OrderBy(c => c.Name);
+                    concurrencyProperties = model.GetConcurrencyProperties(entityInstanceContext.NavigationSource).OrderBy(c => c.Name);
```
Comment (robward-ms) [inline OData/src/System.Web.OData/OData/ETagMessageHandler.cs]: >     [](start = 16, length = 4)

nit: over-indented.

File: OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs [hunk 2]
```diff
@@ -44,9 +46,12 @@ public static IEdmModel BuildEdmModel(ODataModelBuilder builder)
 
             // Build the navigation source map
             IDictionary<string, EdmNavigationSource> navigationSourceMap = model.GetNavigationSourceMap(builder, edmTypeMap, navigationSources);
-
+
```
Comment (robward-ms) [inline OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs]: nit: extra white space.

File: OData/test/E2ETest/WebStack.QA.Test.OData/ETags/PatchUpdatedEntryWithIfMatchETagsTest.cs [hunk 3]
```diff
@@ -28,56 +30,74 @@ public static void UpdateConfiguration(HttpConfiguration configuration)
...
                     routeName: "odata",
                     routePrefix: "odata",
                     model: model,
                     pathHandler: new DefaultODataPathHandler(),
-                    routingConventions: ODataRoutingConventions.CreateDefault(),
+                    routingConventions: ODataRoutingConventions.CreateDefaultWithAttributeRouting(configuration,model),
                     defaultHandler: HttpClientFactory.CreatePipeline(
                         innerHandler: new HttpControllerDispatcher(configuration),
                         handlers: new[] { new System.Web.OData.ETagMessageHandler() }));
+            //configuration.
```
Comment (robward-ms) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ETags/PatchUpdatedEntryWithIfMatchETagsTest.cs]: >configuration [](start = 14, length = 13)

delete?

File: OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs [hunk 4]
```diff
@@ -548,7 +564,15 @@ private static void AddCoreVocabularyAnnotations(this EdmModel model, Navigation
 
             if (edmProperties.Any())
             {
-                model.SetOptimisticConcurrencyAnnotation(target, edmProperties);
+                // todo: fix SetOptimisticConcurrencyAnnotation to support setting concurrency annotations on singletons
```
Comment (robward-ms) [inline OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs]: >todo [](start = 19, length = 4)

file a github issue?

**Comment by robward-ms** [comment]
:shipit:

File: OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs (hunk 4 above)
Comment (mikepizzo) [inline OData/src/System.Web.OData/OData/Builder/EdmModelHelperMethods.cs]: Created [https://github.com/OData/odata.net/issues/770](https://github.com/OData/odata.net/issues/770).

---
In reply to: [105059250](https://github.com/OData/WebApi/pull/931#discussion_r105059250) [](ancestors = 105059250)

**Pull request merged**